llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,455 @@
+
+      BlockedPhysRegs.push_back(PhysReg);
+    }
+  }
+
+  LLVM_DEBUG(dbgs() << "Forbidden Physical Register of the bb."
+                    << MBB->getNumber() << "\n");
+  for (auto &Reg : BlockedPhysRegs) {
+    LLVM_DEBUG(dbgs() << printReg(Reg, TRI, 0, &MRI) << "\n");
+  }
+
+  return BlockedPhysRegs;
+}
+
+bool replaceReg(const Register Reg,
+                const SmallVector<MCPhysReg, 16> &ForbiddenPhysRegs,
+                SmallVector<MCPhysReg, 16> &AlreadyReplaced,
+                const MachineFunction &MF, VirtRegMap &VRM, LiveRegMatrix &LRM,
+                const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+                const AIEBaseRegisterInfo &TRI) {
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI.getName(VRM.getPhys(Reg)) << "\n");
+  SmallVector<MCPhysReg, 16> Replacements = getPriorityReplacementRegs(
+      Reg, ForbiddenPhysRegs, MF, VRM, LRM, MRI, LIS, TRI);
+
+  for (const MCPhysReg &PhysReg : Replacements) {
+    if (find(AlreadyReplaced, PhysReg) != AlreadyReplaced.end())
+      continue;
+
+    const LiveInterval &LI = LIS.getInterval(Reg);
+    LiveRegMatrix::InterferenceKind IK = LRM.checkInterference(LI, PhysReg);
+
+    if (IK == LiveRegMatrix::IK_Free) {
+      LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                        << Register::virtReg2Index(Reg) << " will replace "
+                        << printReg(VRM.getPhys(Reg), &TRI, 0, &MRI) << " with "
+                        << printReg(PhysReg, &TRI, 0, &MRI) << '\n');
+
+      VRM.clearVirt(Reg);
+      LRM.assign(LI, PhysReg);
+      AlreadyReplaced.push_back(PhysReg);
+      return true;
+    }
+  }
+  return false;
+}
+
+SmallVector<MCPhysReg, 16> getPriorityReplacementRegs(
+    const Register Reg, const SmallVector<MCPhysReg, 16> &ForbiddenPhysRegs,
+    const MachineFunction &MF, const VirtRegMap &VRM, LiveRegMatrix &LRM,
+    const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+    const AIEBaseRegisterInfo &TRI) {
+  SmallVector<MCPhysReg, 16> PhysRegs;
+  const TargetRegisterClass *RC = MRI.getRegClass(Reg);
+
+  for (const MCPhysReg &PhysReg : RC->getRegisters()) {
+    // iterate Regs of RC and allocate first found reg that does not
+    // interfere
+    if (find(ForbiddenPhysRegs, PhysReg) != ForbiddenPhysRegs.end())
+      continue;
+
+    LiveRegMatrix::InterferenceKind IK =
+        LRM.checkInterference(LIS.getInterval(Reg), PhysReg);
+    if (IK == LiveRegMatrix::IK_Free)
+      PhysRegs.push_back(PhysReg);
+  }
+
+  std::sort(PhysRegs.begin(), PhysRegs.end(), std::less<MCPhysReg>());
+
+  LLVM_DEBUG(dbgs() << "Renaming Candidate " << printReg(Reg, &TRI, 0, &MRI)
+                    << " Ordering";
+             for (const MCPhysReg &PhysReg
+                  : PhysRegs) {
+               dbgs() << " " << printReg(PhysReg, &TRI, 0, &MRI);
+             } dbgs()
+             << "\n";);
+
+  return PhysRegs;
+}
+
+static bool isUndefReserved(
```

Comment:
I'd be happy with a docstring :) It's not entirely clear to me what we are verifying, and for what reason

Reply:
added documentation